core/unix/signal.c

```diff
@@ -6015,8 +6019,16 @@ handle_sigreturn(dcontext_t *dcontext, void *ucxt_param, int style)
          * when itself was non-rt?
          */
 
-        /* Discard blocked signals, re-set from prev mask stored in frame. */
-        set_blocked(dcontext, SIGMASK_FROM_UCXT(ucxt), true /*absolute*/);
+        if (info->pre_syscall_app_sigprocmask_valid) {
+            /* Discard blocked signals, re-set from original sigmask we have changed
+             * in pre_system_call of extended p* version of a system call.
+             */
+            set_blocked(dcontext, &info->pre_syscall_app_sigprocmask, true /*absolute*/);
+            info->pre_syscall_app_sigprocmask_valid = false;
```

Comment:
I don't follow the high-level reason for doing this: the comment does not seem to contain enough context as to the assumptions and goal here.

The app does not have to invoke sigreturn out of a handler: it can and does use longjmp (hopefully siglongjmp).  DR should not rely on coming here to reset any state.

Xref my comments above in pre-syscall about restoring the mask: are you saying the mask restore needs to be delayed?  Please explain that in the pre-syscall comments.  If it's about getting the mask right for the handler, we may need to restore this p* syscall state right after setting up the handler's signal frame and to-restore mask inside the frame.

Reply:
> The app does not have to invoke sigreturn out of a handler
I agree to that. It turned out to be difficult to find the right place where to restore the mask we had modified pre_syscall. The dispatcher doesn't call post_syscall if pre_syscall returned EINTR error (which is convenient so the syscall doesn't get executed).
> re you saying the mask restore needs to be delayed
Yes, if we restore too early, we won't receive pending signals.
> If it's about getting the mask right for the handler, we may need to restore this p* syscall state right after setting up the handler's signal frame and to-restore mask inside the frame.
Where are you suggesting that should happen?

Reply:
Done. Mask is not restored when setting up app's signal frame.